### PR TITLE
Refactor viz code + transform keypoints + integrated keypoint viz code

### PR DESCRIPTION
**Pull request by ElmoPA**

### CODE REVIEW

**Comment by ElmoPA** [comment]
* **#274** <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/274?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#275](https://github.com/GaTech-RL2/EgoVerse/pull/275) <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/275?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#278](https://github.com/GaTech-RL2/EgoVerse/pull/278) <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/278?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#273** <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/273?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/273?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: egomimic/scripts/zarr_data_viz.ipynb [hunk 1]
```diff
@@ -0,0 +1,420 @@
+{
```
Comment (SimarKareer) [inline egomimic/scripts/zarr_data_viz.ipynb:1]: Delete this file

File: egomimic/scripts/tutorials/zarr_data_viz.ipynb [hunk 2]
```diff
@@ -1,215 +1,295 @@
...
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "id": "a4aa1a05",
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "key_map = Eva.get_keymap()\n",
+                "transform_list = Eva.get_transform_list(mode=\"cartesian_wristframe\")\n",
```
Comment (SimarKareer) [inline egomimic/scripts/tutorials/zarr_data_viz.ipynb]: Default should be cartesian_camframe

File: egomimic/rldb/embodiment/human.py [hunk 3]
```diff
@@ -103,7 +296,428 @@ class Mecka(Human):
...
+        right_wrist_action_headframe,
+    ]
+    if delete_target_world:
+        delete_keys.append(target_world)
+        if target_world_is_quat:
+            delete_keys.append(target_world_ypr)
+    transform_list.extend(
+        [
+            Reshape(
+                input_key=left_keypoints_action_headframe,
```
Comment (SimarKareer) [inline egomimic/rldb/embodiment/human.py:426]: Can you allow reshape to take in a list of reshapes to do. smth like  
Reshape(  
[(input_key, output_key, shape), (input_key2, output_key2, shape2)]  
)

File: egomimic/rldb/embodiment/human.py [hunk 4]
```diff
@@ -22,75 +31,259 @@ class Human(Embodiment):
...
+        return ims
 
     @classmethod
-    def viz_transformed_batch(cls, batch, mode=""):
-        image_key = cls.VIZ_IMAGE_KEY
-        action_key = "actions_cartesian"
+    def viz_transformed_batch(
+        cls,
+        batch,
+        mode=Literal["traj", "axes", "keypoints"],
```
Comment (SimarKareer) [inline egomimic/rldb/embodiment/human.py:63]: what is this literal thing?

File: egomimic/rldb/embodiment/human.py [hunk 5]
```diff
@@ -22,75 +31,259 @@ class Human(Embodiment):
...
-            return _viz_batch_palm_traj(
-                batch=batch,
-                image_key=image_key,
-                action_key=action_key,
+    @classmethod
+    def viz(
+        cls,
+        images,
+        actions,
+        mode=Literal["traj", "axes", "keypoints"],
```
Comment (SimarKareer) [inline egomimic/rldb/embodiment/human.py:92]: same here

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 6]
```diff
@@ -167,7 +161,8 @@ def __init__(
         self.chunk_world = chunk_world
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py:161]: <p class='graphite__hidden'><i>[Re: line +158]</i></p>
add docstring for mode and inverse
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/273?utm_source=unchanged-line-comment">Graphite</a>.</p>

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 7]
```diff
@@ -183,25 +178,56 @@ def transform(self, batch):
...
+            to_matrix_fn = _xyzwxyz_to_matrix
+        elif self.mode == "xyzypr":
+            to_matrix_fn = _xyzypr_to_matrix
+        elif self.mode == "xyz":
+            to_matrix_fn = _xyz_to_matrix
+        else:
+            raise ValueError(f"Invalid mode: {self.mode}")
 
+        target_world_to_matrix_fn = (
+            _xyzwxyz_to_matrix if target_world.shape[-1] == 7 else _xyzypr_to_matrix
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py:203]: Why does this rely on explicitly checking dim 7

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 8]
```diff
@@ -390,6 +410,17 @@ def transform(self, batch):
 # ---------------------------------------------------------------------------
 # Shape Transforms
 # ---------------------------------------------------------------------------
+class SplitKeys(Transform):
+    def __init__(self, input_key: str, output_key_list: list[(str, int)]):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py:414]: Write a docstring, this is a weird transform

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 9]
```diff
@@ -414,10 +445,23 @@ def transform(self, batch):
 
         return batch
 
+
+class Reshape(Transform):
+    def __init__(self, input_key: str, output_key: str, shape: tuple):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py:450]: Docstring

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 10]
```diff
@@ -429,5 +473,7 @@ def transform(self, batch: dict) -> dict:
             elif isinstance(batch[key], torch.Tensor):
                 batch[key] = batch[key].clone()
             else:
-                raise ValueError(f"NumpyToTensor expects key '{key}' to be a numpy array or torch tensor, got {type(batch[key])}")
+                raise ValueError(
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py:476]: fix formatting diff, did you run ruff in precommit?

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 11]
```diff
@@ -716,12 +716,11 @@ def __getitem__(self, idx: int) -> dict[str, torch.Tensor]:
         if self.transform:
             for transform in self.transform or []:
                 try:
+                    # breakpoint()
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py:741]: rm breakpoint

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 12]
```diff
@@ -716,12 +716,11 @@ def __getitem__(self, idx: int) -> dict[str, torch.Tensor]:
...
                     data = transform.transform(data)
                 except Exception as e:
                     logger.error(f"Error transforming data: {e}")
-                    logger.error(f"Data: {data}")
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py:744]: Why'd you get rid of this?

File: egomimic/scripts/tutorials/zarr_data_viz.ipynb [hunk 13]
```diff
@@ -1,215 +1,295 @@
...
+            "id": "79d184b3",
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "%load_ext autoreload\n",
+                "%autoreload 2"
+            ]
+        },
+        {
+            "cell_type": "markdown",
```
Comment (SimarKareer) [inline egomimic/scripts/tutorials/zarr_data_viz.ipynb]: what's causing this large diff?

File: egomimic/scripts/tutorials/zarr_data_viz.ipynb [hunk 14]
```diff
@@ -258,7 +258,7 @@
  ],
```
Comment (SimarKareer) [inline egomimic/scripts/tutorials/zarr_data_viz.ipynb:258]: <p class='graphite__hidden'><i>[Re: line +2]</i></p>
Run this notebook again and make sure it runs, I got an error
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/273?utm_source=unchanged-line-comment">Graphite</a>.</p>

File: egomimic/scripts/zarr_data_viz.ipynb [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+    "\n",
+    "cloudflare_ds = MultiDataset._from_resolver(\n",
+    "    resolver, filters=filters, sync_from_s3=True, mode=\"total\"\n",
+    ")\n",
+    "\n",
+    "loader_kp = torch.utils.data.DataLoader(cloudflare_ds, batch_size=1, shuffle=False)"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (SimarKareer) [inline egomimic/scripts/zarr_data_viz.ipynb:259]: Update the keypoint viz in this notebook to use your new functions

File: egomimic/scripts/zarr_data_viz.ipynb [hunk 16]
```diff
@@ -0,0 +1,420 @@
+{
+ "cells": [
```
Comment (SimarKareer) [inline egomimic/scripts/zarr_data_viz.ipynb:2]: 1. Make sure this notebook runs
2. Update it to use your new viz code
